d2/src/main/java/com/linkedin/d2/balancer/properties/ServiceStoreProperties.java

```diff
@@ -34,6 +35,7 @@
  * certain objects are serialized differently than how Jackson would serialize the object (for instance, using different key names), and
  * that will cause problems in serialization/deserialization.
  */
+@JsonInclude(JsonInclude.Include.NON_NULL) // NOTE: fields with null values will NOT be serialized (won't be included in ZK data)
```

Comment:
I tried to put the annotation just on the canary fields, but it doesn't work (the canary fields with null values are still serialized).

I checked that all fields in the cluster/service config have default values (empty map, list, etc.), so it won't be affected, (otherwise the tests which verify the values in d2-config-service will fail, please d2-config-service RB). Failout configs don't need the null-value key to be present, confirmed with Shangcheng and Chris.

In general, having null values meaning something is a bad design that we should avoid anyway in future feature designs. It's good to have either default dummy values or not including the key at all.